### ISSUE TITLE
We reached the witch-hunting phase

### ISSUE THREAD

**Issue by algunion** [jinyus/related_post_gen#203]
After the discussion [here ](https://github.com/jinyus/related_post_gen/issues/194), I think it starts to be clear that the _unsafe_ prohibition from any user library is going to be a difficult thing to uphold while also ensuring fairness.

I see a few challenges here:

- The developer challenge: is everybody willing to provide an implementation here under the obligation to invest a tremendous amount of time and check every nook and cranny of the packages he uses so a case can be made that there is no _usafe_ code?
- The watch over the fence issue: I don't like that Rust is fast. I feel that I have reached the limits of my language, so let's take a peak then - Rust uses `bumpalo`, which contains unsafe code: owner, please ban `bumpalo`. To some extent, it seems that some developers are concerned with going over the source code of the packages used by the languages which happen to be at the top. So now we have the non-user of language X reading the code implemented in language X and opening an issue. But then some experienced users and even language developers of X intervene and present an expertise-backed opinion, and they can be easily dismissed (because they are biased and it is natural to take the side of their language).
- The owner's challenge: the owner of the repository will not be able to invoke expertise in all the used languages and ensure a fair judgment over all these issues.  

So where is this going? Are we starting a `rust + bumpalo` round now?  

**Maybe an objective solution is to call this the battle of STDLIBS and prohibit any usage of any external package (while also prohibiting any _unsafe_ call from stdlib).** 

But if we do the above, then this benchmark is no longer real-world (because who would avoid battle-tested packages and start implementing functionality from scratch)?

So how about an even better one: prohibit the developers from doing _unsafe_ tricks, and let's call it a **languages ecosystem benchmark**.

**Comment by DrBlury** [comment]
I think it's fair to assume that IF you use an external package like bumpalo you'd make sure it complies with the rules of the benchmark. If you feel like you've reached the edge of what's possible in your language, that's fine. Maybe someone else can find a neat trick to speed things up. (After all that's what this repo is for!) 

Unsafe code or c bindings in other languages are forbidden as they do not represent the true spirit of the language to benchmark. So in this - I agree with you. Please remove bumpalo from rust as it's against the rules.

I think it's fair to assume that others will watch over each other so noone is using dirty tricks. This means everyone can participate and contribute while others can check their implementation and maybe adjust their own... Or find that you used something that's against the rules (like you with bumpalo for rust)

Using only STDLIBS is not an option, as some languages (CPP for example) do not handle Json decode/encode at all and there is some additional code required.

So I think using packages is fine, as long as the compute step does not use an extremely specific packages or custom type implementations (that's just my feeling) to be faster. Others will check your implementation if you beat them in the benchmark and will make sure you are complying with the rules. 

Unsafe tricks are against the rules anyway. 

That's just my two cents. :)

**Comment by Seelengrab** [comment]
> The watch over the fence issue: I don't like that Rust is fast. I feel that I have reached the limits of my language, so let's take a peak then - Rust uses bumpalo, which contains unsafe code: owner, please ban bumpalo. 

To be clear, I have no issue per se with the Rust version using bumpalo. It's a good crate, and it's fantastic that you can leverage it to gain performance. I do however think it's a bit weird to seemingly arbitrarily apply some "unsafeness" criterion only to some languages - so what I'm asking about is consistency between languages :shrug:

**Comment by michaelsbradleyjr** [comment]
> Unsafe code or c bindings in other languages are forbidden as they do not represent the true spirit of the language to benchmark. So in this - I agree with you. Please remove bumpalo from rust as it's against the rules.

Judgment calls on this can start to be difficult.

For example, the Nim impl for this benchmark [uses the `xxhash` package](https://github.com/jinyus/related_post_gen/blob/main/nim/src/related.nim#L3), and [that package](https://nimble.directory/pkg/xxhash) is a thin [`{.importc.}` wrapper](https://github.com/OpenSystemsLab/xxhash.nim/blob/master/src/xxhash.nim) around the C library [xxHash](https://github.com/Cyan4973/xxHash). The motivation for for using xxHash is simple: Nim's built-in [`murmurHash`](https://github.com/nim-lang/Nim/blob/v2.0.0/lib/pure/hashes.nim#L306) is a bit pokey, the community knows it, and it's [*really easy*](https://github.com/jinyus/related_post_gen/blob/main/nim/src/related.nim#L18-L19) to substitute a faster hash for your HashMaps and HashSets without waiting on `std` to be overhauled, which may never happen.

With Nim in general, this is *the way* of things for many real world projects, and Nim's design promotes and encourages it: use the built-ins and `std` library where they're well suited, and where they're not suited *maybe* you'll implement some alts in your own pure Nim but *most probably* you'll wrap one or more C/++ libraries and/or use wrappers that are already available. With Nim's C backend — it's the default and what most projects use — FFI is extremely thin: there's no bridge really, it's just a matter of the correct symbols being available when the Nim compiler hands off the C sources it generated to `gcc` or `clang` or whatever.

Summing up, I agree with this
> So how about an even better one: prohibit the developers from doing unsafe tricks, and let's call it a languages ecosystem benchmark.

In the case of Nim specifically, prohibiting the use of wrappers for C/++ libs would be completely against the grain of how the language is designed to be used and is used in practice.
